near-sdk/src/environment/sys.rs

```diff
@@ -0,0 +1,492 @@
+        method_name_len: u64,
+        method_name_ptr: u64,
+        arguments_len: u64,
+        arguments_ptr: u64,
+        amount_ptr: u64,
+        gas: u64,
+    ) -> u64;
+    pub fn promise_and(promise_idx_ptr: u64, promise_idx_count: u64) -> u64;
+    pub fn promise_batch_create(account_id_len: u64, account_id_ptr: u64) -> u64;
+    pub fn promise_batch_then(promise_index: u64, account_id_len: u64, account_id_ptr: u64) -> u64;
+    // #######################
+    // # Promise API actions #
+    // #######################
+    pub fn promise_batch_action_create_account(promise_index: u64);
+    pub fn promise_batch_action_deploy_contract(promise_index: u64, code_len: u64, code_ptr: u64);
+    pub fn promise_batch_action_function_call(
+        promise_index: u64,
+        method_name_len: u64,
+        method_name_ptr: u64,
+        arguments_len: u64,
+        arguments_ptr: u64,
+        amount_ptr: u64,
+        gas: u64,
+    );
+    pub fn promise_batch_action_transfer(promise_index: u64, amount_ptr: u64);
+    pub fn promise_batch_action_stake(
+        promise_index: u64,
+        amount_ptr: u64,
+        public_key_len: u64,
+        public_key_ptr: u64,
+    );
+    pub fn promise_batch_action_add_key_with_full_access(
+        promise_index: u64,
+        public_key_len: u64,
+        public_key_ptr: u64,
+        nonce: u64,
+    );
+    pub fn promise_batch_action_add_key_with_function_call(
+        promise_index: u64,
+        public_key_len: u64,
+        public_key_ptr: u64,
+        nonce: u64,
+        allowance_ptr: u64,
+        receiver_id_len: u64,
+        receiver_id_ptr: u64,
+        method_names_len: u64,
+        method_names_ptr: u64,
+    );
+    pub fn promise_batch_action_delete_key(
+        promise_index: u64,
+        public_key_len: u64,
+        public_key_ptr: u64,
+    );
+    pub fn promise_batch_action_delete_account(
+        promise_index: u64,
+        beneficiary_id_len: u64,
+        beneficiary_id_ptr: u64,
+    );
+    // #######################
+    // # Promise API results #
+    // #######################
+    pub fn promise_results_count() -> u64;
+    pub fn promise_result(result_idx: u64, register_id: u64) -> u64;
+    pub fn promise_return(promise_id: u64);
+    // ###############
+    // # Storage API #
+    // ###############
+    pub fn storage_write(
+        key_len: u64,
+        key_ptr: u64,
+        value_len: u64,
+        value_ptr: u64,
+        register_id: u64,
+    ) -> u64;
+    pub fn storage_read(key_len: u64, key_ptr: u64, register_id: u64) -> u64;
+    pub fn storage_remove(key_len: u64, key_ptr: u64, register_id: u64) -> u64;
+    pub fn storage_has_key(key_len: u64, key_ptr: u64) -> u64;
+    // ###############
+    // # Validator API #
+    // ###############
+    pub fn validator_stake(account_id_len: u64, account_id_ptr: u64, stake_ptr: u64);
+    pub fn validator_total_stake(stake_ptr: u64);
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+mod mock_chain {
```

Comment:
Perhaps we can move this to the existing `mocked_blockchain.rs` file? I think there's a benefit in having `sys.rs` containing *only* the `extern "C"` block.

Reply:
Yeah, I'll do that! I was deciding between the two, but I put it there because I figured if any API changed, it would be easy to see that this mock implementation needs to be updated.